File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    n = len(maps)
    m = len(maps[0])
    
    # 방문 체크 배열
    visited = [[False] * m for _ in range(n)]
    
    # 상하좌우 방향
    directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]
    
    # 큐 생성 및 시작점 추가
    queue = deque()
    queue.append((0, 0, 1))
    visited[0][0] = True  # 시작점 방문 체크
    
    while queue:
        row, col, distance = queue.popleft()
        
        # 목적지 도착 체크
        if row == n-1 and col == m-1:
            return distance
        
        # 4방향 탐색
        for dr, dc in directions:
            new_row = row + dr
            new_col = col + dc
            
            # 이동 가능한지 체크
            if (0 <= new_row < n and 0 <= new_col < m and
                maps[new_row][new_col] == 1 and
                not visited[new_row][new_col]):
                
                visited[new_row][new_col] = True  # 방문 체크
                queue.append((new_row, new_col, distance + 1))    # 큐에 추가
    
    return -1

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cs
using System;
using System.Collections.Generic;

class Solution {
    public int solution(int[,] maps) {
        int row = maps.GetLength(0);
        int col = maps.GetLength(1);
        int[,] dis = new int[row, col];
        Queue<(int x, int y)> que = new Queue<(int, int)>();
        
        int[] dx = {-1, 1, 0, 0};
        int[] dy = {0, 0, -1, 1};
        dis[0,0] = 1;
        que.Enqueue((0,0));
        
        while(que.Count>0){
            var(x,y) = que.Dequeue();
            
            for(int i = 0; i<4; i++){
                int nx = x +dx[i];
                int ny = y +dy[i];
                //맵 밖인가?
                if(nx < 0 || nx>= row || ny < 0 || ny >= col) continue;
                // 벽인가?
                if(maps[nx, ny] == 0) continue;
                // 이미 방문 했나?
                if(dis[nx, ny] != 0) continue;
                
                dis[nx, ny] = dis[x,y] + 1;
                que.Enqueue((nx, ny));
            }
        }
        int answer = dis[row-1, col-1];
        return answer == 0 ? -1 : answer;
    }
}